libc runtime

haampie opened the pull request:
TODO:

- [x] fix some bug in the concretizer that prevents this
- [x] `libc` automatic external
- [x] make `libc` a virtual, make `gcc-runtime` depend on `libc`
- [x] drop `os` match rules on `linux`
- [x] allow reuse of older libc assuming abi compat
- [x] Ensure every compiler we support on linux can detect libc
- [x] Add unit tests stressing libc compatibility
- [x] Ensure externals on `linux` still work correctly
- [x] postponed: deal with out-of-prefix `PT_INTERP` loader path relocation

haampie committed:
Compiler.default_libc

Some logic to detect what libc the c / cxx compilers use by default,
based on `-dynamic-linker`.

The function `compiler.default_libc()` returns a `Spec` of the form
`glibc@x.y` or `musl@x.y` with the `external_path` property set.

The idea is this can be injected as a dependency.

If we can't run the dynamic linker directly, fall back to `ldd` relative
to the prefix computed from `ld.so.`

alalazo commented:
With respect to:

> fix some bug in the concretizer that prevents this (circular dep with gcc-runtime?)

This is working, provided we add / inject `glibc` as an external package. Otherwise there will be circular dependencies due to `glibc` dependencies on `gcc-runtime` and `glibc`. 

For instance, on:

* **Spack:** 0.22.0.dev0 (9a17670e61fc283335208385e11096b95d5e7e67)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

and using:
```yaml
packages:
  glibc:                                                                                                                                                                                                                                                                         
    externals:                                                                                                                                                                                                                                                                   
    - spec: "glibc@=2.31"                                                                                                                                                                                                                                                        
      prefix: /usr          
```

I get the following for:
```console
$ spack solve zlib-ng
```

![zlib](https://github.com/spack/spack/assets/4199709/f87ffcd0-d1e9-46ed-add0-acc38eaa89ad)

haampie commented:
Right, my original idea was to make it an automatic external

alalazo committed:
Add libc dependency to compiled packages and runtime deps

This commit differentiate linux from other platforms by
using libc compatibility as a criterion for deciding
which buildcaches / binaries can be reused. Other
platforms still use OS compatibility.

On linux a libc is injected by all compilers as an implicit
external, and the compatibility criterion is that a libc is
compatible with all other libcs with the same name and a
version that is lesser or equal.

Some concretization unit tests use libc when run on linux.

alalazo commented:
The last commit (https://github.com/spack/spack/pull/43190/commits/19d2c2c6f70d83d0420ed30ef813ac4f7df9214c) gives a fully working GCC, but we still need some tightening for other compilers. Anyhow, to try the new feature:
```console
$ spack mirror add devtools https://raw.githubusercontent.com/alalazo/test-buildcaches/main/devtools
$ spack buildcache keys -it 
$ spack env activate --temp
$ spack add hdf5~mpi
$ spack concretize --reuse -f
```
On my platform:

* **Spack:** 0.22.0.dev0 (19d2c2c6f70d83d0420ed30ef813ac4f7df9214c)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

This reuses a bunch of nodes from the buildcache that has `almalinux8` builds:

![Screenshot from 2024-04-16 18-52-20](https://github.com/spack/spack/assets/4199709/16d07ef9-0482-4c68-afac-69dfa0c714c2)

alalazo commented:
The shell tests is currently failing, because at the moment they contain some mock installation that do not inject libc - while I started requiring that on linux in https://github.com/spack/spack/pull/43190/commits/ebb113ae3c6b760e62ed87cfe5b14a7f10676d3c

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2478,6 +2478,23 @@ def define_runtime_constraints(self):
+            if not compiler.available:
+                continue
+
+            if spack.platforms.host().name == "linux":
```
We may also include BSD (later?). Plus at some point this should be a target not host condition :thinking: 

Can this branch already be written in terms of the platform `compiler` targets?

alalazo replied:
Would you be fine with https://github.com/spack/spack/pull/43190/commits/85d8bb10705c613c5e3ca1ffbcde3bbc74c18488 for now (which just adds a function where to encode when libc compatibility is used)?

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2427,6 +2478,23 @@ def define_runtime_constraints(self):
+                libc = compiler.compiler_obj.default_libc()
+                if libc:
+                    recorder("*").depends_on(
+                        "libc", when=f"%{str(compiler.spec)}", type="link", description="Add libc"
```
```suggestion
                        "libc", when=f"%{compiler.spec}", type="link", description="Add libc"
```

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2427,6 +2478,23 @@ def define_runtime_constraints(self):
+                    )
+                    recorder("*").depends_on(
+                        str(libc),
+                        when=f"%{str(compiler.spec)}",
```
```suggestion
                        when=f"%{compiler.spec}",
```

haampie commented:
```
$ spack spec -L zlib-ng
Input spec
--------------------------------
 -   zlib-ng

Concretized
--------------------------------
 -   gaxnrttfo2lk7yzdd2nsm5yxf55zadsv  zlib-ng@2.1.6%gcc@13.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu23.10-zen2
[+]  tofiopdeqtfqrjqrfnclgnvx2d47yraq      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen2
[e]  gx2zlxihjlhzfkcz47rjr2iir5hrogi7          ^glibc@2.38%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen2
[e]  gx2zlxihjlhzfkcz47rjr2iir5hrogi7      ^glibc@2.38%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen2
[+]  boxbsmc7a5ycptkstw777ao5xecdo34h      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-ubuntu23.10-zen2
```

not sure what that is about

alalazo commented:
> not sure what that is about.

The issue was that the libc was not "reused". Fixed in https://github.com/spack/spack/pull/43190/commits/b049375976d012f110becb78824a5df63edaaedc

haampie commented:
To sum up the relocation struggles:

1. The interpreter path is on most distros "out of prefix", meaning default prefix -> prefix relocation is insufficient
2. On almost all distros it's out of prefix exactly such that relocation is unnecessary
3. *But* we do care about that small fraction of distros that is different (EESSI users w/ Gentoo Prefix, and NixOS)
4. We preferably don't wanna model in terms of *the* system libc, there can be multiple
5. It's most robust to determine the interpreter path from the compiler's link line
6. But we also want `spack install gcc` to work without a compiler installed: there may not be a compiler to query where libc is.
7. If no configured compiler: we can determine the interpreter path from the current Python process's `PT_INTERP` (except if Python is static, which is a valid use case when bootstrapping Spack on systems without Python)
8. Alternatively we could run `/usr/bin/ldd` or `ldd` from PATH, or scan `/usr` for a `libc.so` with arch compatible with the currently running Python process.

One way or another we get the interpreter path, and add a replacement `<old interpreter path> -> <new interpreter path>`.

alalazo commented:
@spackbot run pipeline

haampie commented:
We decided to stick with exact prefix match for now so that reuse between nix / gentoo prefix and classical distros isn't possible. That makes the scope of this pr a bit smaller.
